### ISSUE TITLE
Save file - filename with "_" in place of " "

### ISSUE THREAD

**Issue by poitierjohan** [spipu/html2pdf#293]
Hi everybody,

When I generate my PDF, all is right.
But, when I save my pdf on my computer, my filename ("Bills january.pdf") is "Bills_january.pdf".

But, I d'ont want this character "_" in place of " ".

How can I resolve this problem ?

Thank your for your following and answers :-)

**Comment by LittleBigFox** [comment]
Hi!

If you write it in a file, you can rename it after save. Il you download it with html2pdf, i think you have to write it first and send it to the browser after.

Blank space are replaced for security reason in many file manager, specialy on the web...

**Comment by spipu** [comment]
sapce in filename is a very very bad practice for web

**Comment by lenamtl** [comment]
Hi, 

You can edit the regex for the name

The file is located in vendor\tecnickcom\tcpdf\tcpdf.php

You can comment the line that is replacing the space by underscore 
or change it to what you prefer could be a dash, no space at all ....
(it is a good practice to replace empty space by something for file name)

```
if ($dest[0] != 'F') {

       // this is removing empty space and replace with underscore
       /* $name = preg_replace('/[\s]+/', '_', $name); */
     
       // this keep letter and number only no accent
     $name = preg_replace('/[^a-zA-Z0-9_\.-]/', '', $name);
}
```

**Comment by spipu** [comment]
never changing the files in the vendor folder.
build directly a good filename before give it to html2pdf !

**Comment by lenamtl** [comment]
I made a test, and even if I build the name first the tcpdf config file will override it...
If you have a working solution, I'm curious to see it as I don't like to edit vendor files.

**Comment by spipu** [comment]
I will not help on this issue, use the good practices

**Comment by lenamtl** [comment]
Hi,

I worked on several Opensource project for several years, one of the goal of Open Source is to help each other so everybody can learn.

I'm using best practice in fact I'm using timestamp for the file name which is the best way to my opinion when saving file on server as each file name are unique using this method and contains only number. 

@poitierjohan ask about a solution for empty space, I provided a solution and explained that is not a good practice, after that @poitierjohan can take the decision to change it or not, but at least know where to change it and why this is happening. 

So now what can we do without changing the tcpdf $name function:
What we can do is to pass a regex on the generated $name directly on the code page prior sending it to the PDF script.
Let say we want to replace accent letter by the corresponding letter or replace space by dash instead underscore, then we have no need to touch the tcpdf file (I have not tested it but this should work).
For example **été** will result t.pdf and **summer 2018** will result summer_2018.pdf with the actual regex.

But if you absolutely need the empty space in the file name (which is not good practice) there is no other way that editing the tcpdf regex.

About Vendor files and folders:
If you absolutely need to change the vendor files / folder ie remove extra fonts (to save considered space) it is a good idea to keep the original package as a backup, document all changes like you do for other script and if later when you update to a new version you will have everything to compare package and made the change if still needed it.

The idea here is to use the best practice and find a way to get the desire result without changing the vendor file when "this is possible", like we do for any other script. When there is no other solution at least to be able to know what have been changed and know the consequence if any.
